lookout/style/format/benchmarks/top_repos_quality.py

```diff
@@ -0,0 +1,349 @@
+        numerator += row[col]
+        denominator += 1
+    if denominator == 0:
+        return 1
+    return numerator / denominator
+
+
+def get_precision_recall_f1_support(report: str) -> (float, float, float, int):
+    """Extract avg / total precision, recall, f1 score, support from report."""
+    for line in report.split("\n"):
+        if "weighted avg" in line:
+            _, prec, recall, f1, support = line.split("` | `")
+            prec, recall, f1, support = [float(f.replace("`", "").replace("|", "")) for f in
+                                              (prec, recall, f1, support)]
+            return prec, recall, f1, int(support)
+
+
+def get_model_summary(report: str) -> (int, float):
+    """Extract model summary - number of rules and avg. len."""
+    pattern = " rules, avg.len. "
+    for line in report.split("\n"):
+        if pattern in line:
+            line = line.replace("`", "")
+            res = list(map(float, line.split(pattern)))
+            assert len(res) == 2
+            n_rules, avg_len = res
+            n_rules = int(n_rules)
+    return n_rules, avg_len
+
+
+def main(args):
+    """Entry point for quality report generation."""
+    slogging.setup(args.log_level, False)
+    if not server.exefile.exists():
+        server.fetch()  # download executable
+    # prepare output directory
+    dir_to_save = args.output
+    os.makedirs(dir_to_save, exist_ok=True)
+    assert os.path.isdir(dir_to_save), "Output should be a directory"
+    reports = []
+    errors_loc = os.path.join(dir_to_save, "errors.txt")
+    port = server.find_port()
+    review_config = {QualityReportAnalyzer.name: {
+        "aggregate": True,
+        }}
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        database = args.database if args.database else os.path.join(tmpdirname, "db.sqlite3")
+        fs = args.fs if args.fs else os.path.join(tmpdirname, "models")
+        os.makedirs(fs, exist_ok=fs)
+        with AnalyzerContextManager(port=port, db=database, fs=fs,
+                                    analyzer="lookout.style.format.quality_report", init=False):
+            for repo in REPOSITORIES:
+                repo, to_commit, from_commit = repo.split()
+                report_loc = os.path.join(dir_to_save, get_repo_name(repo))
+                quality_rep_loc = report_loc + ".quality_report.md"
+                model_rep_loc = report_loc + ".model_report.md"
+                # generate or read report
+                try:
+                    if args.force or not os.path.exists(quality_rep_loc) or \
+                            not os.path.exists(model_rep_loc):
+                        # Skip this step if report was already generated
+                        quality_report, model_report = measure_quality(
+                            repo, to_commit=to_commit, from_commit=from_commit, port=port,
+                            review_config=review_config)
+                        if quality_report is not None:
```

Comment:
BIG LARGE GIANT TODO: convert this horror to Jinja ASAP

Reply:
What do you mean? Jinja is inside `generate_report` and `genearte_model_report` function. here we just concatenate the reports.

Reply:
There are tons of custom string manipulation here

Reply:
I see only this one https://github.com/src-d/style-analyzer/blob/c10c16a4bda15d880c16ee8f7659f8856966c573/lookout/style/format/benchmarks/top_repos_quality.py#L295

All the rest is just writing to the file. is anything missing from my eye?

Reply:
Actually, all the rest :laughing: Everything is just wrong: `quality_report`, `model_report`, etc.

Reply:
So, let be more precise with todos: 
Should I put them to `quality_report` and  `model_report` ?